homeassistant/components/mqtt/mixins.py

```diff
@@ -0,0 +1,471 @@
+            elif self._discovery_update:
+                if old_payload != self._discovery_data[ATTR_DISCOVERY_PAYLOAD]:
+                    # Non-empty, changed payload: Notify component
+                    _LOGGER.info("Updating component: %s", self.entity_id)
+                    await self._discovery_update(payload)
+                else:
+                    # Non-empty, unchanged payload: Ignore to avoid changing states
+                    _LOGGER.info("Ignoring unchanged update for: %s", self.entity_id)
+            async_dispatcher_send(
+                self.hass, MQTT_DISCOVERY_DONE.format(discovery_hash), None
+            )
+
+        if discovery_hash:
+            debug_info.add_entity_discovery_data(
+                self.hass, self._discovery_data, self.entity_id
+            )
+            # Set in case the entity has been removed and is re-added, for example when changing entity_id
+            set_discovery_hash(self.hass, discovery_hash)
+            self._remove_signal = async_dispatcher_connect(
+                self.hass,
+                MQTT_DISCOVERY_UPDATED.format(discovery_hash),
+                discovery_callback,
+            )
+            async_dispatcher_send(
+                self.hass, MQTT_DISCOVERY_DONE.format(discovery_hash), None
+            )
+
+    async def async_removed_from_registry(self) -> None:
+        """Clear retained discovery topic in broker."""
+        if not self._removed_from_hass:
+            discovery_topic = self._discovery_data[ATTR_DISCOVERY_TOPIC]
+            publish(self.hass, discovery_topic, "", retain=True)
+
+    @callback
+    def add_to_platform_abort(self) -> None:
+        """Abort adding an entity to a platform."""
+        if self._discovery_data:
+            discovery_hash = self._discovery_data[ATTR_DISCOVERY_HASH]
+            clear_discovery_hash(self.hass, discovery_hash)
+            async_dispatcher_send(
+                self.hass, MQTT_DISCOVERY_DONE.format(discovery_hash), None
+            )
+        super().add_to_platform_abort()
+
+    async def async_will_remove_from_hass(self) -> None:
+        """Stop listening to signal and cleanup discovery data.."""
+        self._cleanup_discovery_on_remove()
+
+    def _cleanup_discovery_on_remove(self) -> None:
+        """Stop listening to signal and cleanup discovery data."""
+        if self._discovery_data and not self._removed_from_hass:
+            debug_info.remove_entity_data(self.hass, self.entity_id)
+            clear_discovery_hash(self.hass, self._discovery_data[ATTR_DISCOVERY_HASH])
+            self._removed_from_hass = True
+
+        if self._remove_signal:
+            self._remove_signal()
+            self._remove_signal = None
+
+
+def device_info_from_config(config):
+    """Return a device description for device registry."""
+    if not config:
+        return None
+
+    info = {
+        "identifiers": {(DOMAIN, id_) for id_ in config[CONF_IDENTIFIERS]},
```

Comment:
could these strings be imported ?

Reply:
No, they're hardcoded strings in `homeassistant/helpers/device_registry.py`